src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,163 @@
-        return null;
+    public T remove(int index) {
+        checkIndexForGetAndSet(index);
+        return unlinkNode(getNodeByIndex(index));
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public boolean remove(T value) {
+        Node<T> nodeForRemove = head;
+        while (nodeForRemove != null) {
+            if (nodeForRemove.elementValue == null
+                    || (nodeForRemove.elementValue != null
+                    && nodeForRemove.elementValue.equals(value))) {
+                unlinkNode(nodeForRemove);
+                return true;
+            }
+            nodeForRemove = nodeForRemove.next;
+        }
+        return false;
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public T set(T value, int index) {
+        checkIndexForGetAndSet(index);
+        Node<T> nodeForChangeValue = getNodeByIndex(index);
+        T oldValue = nodeForChangeValue.elementValue;
+        nodeForChangeValue.elementValue = value;
+        return oldValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addToTail(T value) {
+        Node<T> lastElement = tail;
+        Node<T> newNode = new Node<>(lastElement, value, null);
+        tail = newNode;
+        if (lastElement != null) {
+            lastElement.next = newNode;
+        } else {
+            head = newNode;
+        }
+        size++;
+    }
+
+    private void checkIndexForGetAndSet(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(MESSAGE_INDEX_ERROR);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(MESSAGE_INDEX_ERROR);
+        }
+    }
+
+    private void insertBefore(T value, Node<T> node) {
+        Node<T> linkToPrevious = node.previous;
+        Node<T> insertedNode = new Node<>(linkToPrevious, value, node);
+        node.previous = insertedNode;
+        if (linkToPrevious == null) {
+            head = insertedNode;
+        } else {
+            linkToPrevious.next = insertedNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> searchNode;
+        if (index <= (size / 2)) {
+            searchNode = head;
+            for (int i = 0; i < index; i++) {
+                searchNode = searchNode.next;
+            }
+        } else {
+            searchNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                searchNode = searchNode.previous;
+            }
+        }
+        return searchNode;
+    }
+
+    private T unlinkNode(Node<T> node) {
+        final Node<T> nextElement = node.next;
+        final Node<T> previousElement = node.previous;
```

Comment:
why final?

Reply:
has a problem: Distance between variable 'element' declaration and its first usage is 4, but allowed 3.  Consider making that variable final if you still need to store its value in advance (before method calls that might have side effects on the original value). [VariableDeclarationUsageDistance]. 
Fixed.